lib/runner/pre.js

```diff
@@ -0,0 +1,189 @@
+                }
+                // stream is valid, so replace the value in data
+                else {
+                    // @todo: why is this not `value` like formdata. it would have made it easy to pass through
+                    // a common logic
+                    filedata.content = stream;
+                }
+
+                next();
+            }
+        }[mode] || async.constant()], function (err) {
+            // just as a precaution, show the error in console. each resolver anyway should handle their own console
+            // warnings.
+            // @todo - get cursor here.
+            err && triggers.console(cursor, 'warn', 'file data resolution error: ' + (err.message || err));
+            done(null); // absorb the error since a console has been trigerred
+        });
+    },
+    // Proxy lookup
+    function (payload, done) {
+        var request = _.get(payload, 'item.request'),
+            url,
+            proxies,
+            proxyConfig;
+
+        if (!request) { return done(new Error('No request to resolve proxy for.')); }
+
+        url = request.url && request.url.toString();
+        proxies = this.options.proxies;
+
+        if (_.isFunction(_.get(proxies, 'resolve'))) {
+            proxyConfig = proxies.resolve(url);
+            proxyConfig && (request.proxy = proxyConfig);
+        }
+        done();
+    },
+    // Certificate lookup + reading from whichever file resolver is provided
+    function (payload, done) {
+        var keyPath,
+            certPath,
+            request,
+            fileResolver,
+
+            certificate;
+
+        // A. Check if we have the file resolver
+        fileResolver = _.get(this.options, 'fileResolver') ||
+            // todo: remove this once apps give the options correctly
+            _.get(this.options, 'requester.fileResolver');
+
+        if (!fileResolver) { return done(); }  // No point going ahead
+
+        // B. Ensure we have the request
+        request = _.get(payload, 'item.request');
+        if (!request) { return done(new Error('No request to resolve proxy for.')); }
+
+        // C. See if any cert should be sent, by performing a URL matching
+        certificate = this.options.certificates && this.options.certificates.resolveOne(request.url);
+        if (!certificate) { return done(); }
+
+        // D. Fetch the paths
+        keyPath = _.get(certificate, 'key.src');
+        certPath = _.get(certificate, 'cert.src');
+
+        // E. Read from the path, and add the values to the certificate, also associate
+        // the certificate with the current request.
+        async.mapValues({
+            key: keyPath,
+            cert: certPath
+        }, function (value, key, cb) {
+            fileResolver.readFile(value, cb);
+        }, function (err, fileContents) {
+            if (err) { return done(err); }
+
+            _.set(certificate, 'key.value', fileContents.key);
```

Comment:
I just have one concern.

This request has complete certificate info, including passphrase and certificate contents(right now `certificate.toJSON()` will pass through the `key.value`). Is there a possibility of this request object persisted by any chance?